staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -861,13 +890,21 @@ func (c *Cacher) GetList(ctx context.Context, key string, opts storage.ListOptio
 	//   the elements in ListObject are Struct type, making slice will bring excessive memory consumption.
 	//   so we try to delay this action as much as possible
 	var selectedObjects []runtime.Object
-	for _, obj := range objs {
+	var lastSelectedObjectKey string
+	var hasMoreListItems bool
+	applyLimit := shouldApplyListLimit(opts)
+	for i, obj := range objs {
 		elem, ok := obj.(*storeElement)
 		if !ok {
 			return fmt.Errorf("non *storeElement returned from storage: %v", obj)
 		}
 		if filter(elem.Key, elem.Labels, elem.Fields) {
 			selectedObjects = append(selectedObjects, elem.Object)
+			lastSelectedObjectKey = elem.Key
+		}
+		if applyLimit && int64(len(selectedObjects)) >= opts.Predicate.Limit {
+			hasMoreListItems = i < len(objs)-1
```

Comment:
alternatively we could loop over the remaining `objs` and see if there is at least one matching item.

Reply:
Let's not overoptimize.